diff --git a/cpuexec.cpp b/cpuexec.cpp
index cd2b1ba..50f078b 100644
--- a/cpuexec.cpp
+++ b/cpuexec.cpp
@@ -16,6 +16,7 @@
 #include "debug.h"
 #include "missing.h"
 #endif
+#include <set>
 
 static inline void S9xReschedule (void);
 
@@ -167,6 +168,34 @@ void S9xMainLoop (void)
 				Opcodes = S9xOpcodesSlow;
 		}
 
+		static std::string filename = "instruction_trace.txt";
+		static std::fstream out( filename, std::ofstream::in | std::ofstream::out | std::ofstream::app );
+		static std::set<uint32> codeAddresses;
+		static boolean loadedExistingAddresses = false;
+
+		if ( !loadedExistingAddresses )
+		{
+			while ( !out.eof() )
+			{
+				uint32 memory_mode = 0;
+				uint32 index_mode = 0;
+				uint32 address = 0;
+
+				out >> memory_mode >> index_mode >> address;
+				codeAddresses.insert( address );
+			}
+
+			out.clear();
+
+			loadedExistingAddresses = true;
+		}
+		auto search = codeAddresses.find( Registers.PBPC );
+		if ( search == codeAddresses.end() )
+		{
+			codeAddresses.insert( Registers.PBPC );
+			out << (CheckMemory() ? 1 : 0) << " " << ( CheckIndex() ? 1 : 0 ) << " " << Registers.PBPC << std::endl;
+		}
+
 		Registers.PCw++;
 		(*Opcodes[Op].S9xOpcode)();
 
